core/coreapi/client/client.go

```diff
@@ -0,0 +1,146 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	config "github.com/ipfs/go-ipfs-config"
+	"github.com/ipfs/go-ipfs-http-client"
+	"github.com/ipfs/interface-go-ipfs-core"
+	"github.com/mitchellh/go-homedir"
+	"github.com/pkg/errors"
+
+	"github.com/ipfs/go-ipfs/core"
+	"github.com/ipfs/go-ipfs/core/coreapi"
+	"github.com/ipfs/go-ipfs/repo/fsrepo"
+)
+
+const (
+	DefaultPathName = ".ipfs"
+	DefaultPathRoot = "~/" + DefaultPathName
+	EnvDir          = "IPFS_PATH"
+)
+
+func repo() (string, error) {
+	baseDir := os.Getenv(EnvDir)
+	if baseDir == "" {
+		baseDir = DefaultPathRoot
+	}
+	return homedir.Expand(baseDir)
+}
+
+//////////////////////
+
+type ManagedApi interface {
+	iface.CoreAPI
+
+	io.Closer
+}
+
+type ctxApi struct {
+	iface.CoreAPI
+	context.CancelFunc
+}
+
+func (a *ctxApi) Close() error {
+	a.CancelFunc()
+	return nil
+}
+
+var _ ManagedApi = &ctxApi{}
+
+//////////////////////
+
+func embedded(repoPath string) (ManagedApi, error) {
+	ctx, cancel := context.WithCancel(context.TODO())
+
+	r, err := fsrepo.Open(repoPath)
+	if err != nil {
+		return nil, fmt.Errorf("opening fsrepo failed: %s", err)
+	}
+	n, err := core.NewNode(ctx, &core.BuildCfg{
+		Online: true,
+		Repo:   r,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("ipfs NewNode() failed: %s", err)
+	}
+
+	api, err := coreapi.NewCoreAPI(n)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ctxApi{
+		CoreAPI: api,
+		CancelFunc: cancel,
+	}, nil
+}
+
+type ApiProvider func() (ManagedApi, error)
+
+var localApi ApiProvider = func() (ManagedApi, error) {
+	api, err := httpapi.NewLocalApi()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ctxApi{
+		CoreAPI: api,
+		CancelFunc: func() {},
+	}, nil
+}
+
+var tempEmbedded ApiProvider = func() (ManagedApi, error) {
+	dir, err := ioutil.TempDir("", "ipfs-shell")
```

Comment:
Ideally, we'd remove this directory (wrap the API so it gets deleted on close?).